workflow-manager/limiter/limiter.go

```diff
@@ -0,0 +1,69 @@
+package limiter
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Limiter is similar to sync.Pool, however, it limtis how many shared
+// objects (in this case workers) there can be
+type Limiter struct {
+	workers    chan int32
+	inProgress int32
+}
+
+// Ticket is the structure that allows you to schedule jobs
+type Ticket struct {
+	limiter *Limiter
+	worker  int32
+	dirty   int32
+}
+
+// New creates a new instance of Limiter that limits how many functions can get executed without the other ones completing first
+func New(maxWorkers int32) *Limiter {
+	l := &Limiter{make(chan int32, maxWorkers), 0}
+
+	for i := int32(0); i < maxWorkers; i++ {
+		l.workers <- i
+	}
+
+	return l
+}
+
+// RequestTicket requests a ticket for a worker
+// this call will block if there are no tickets available for work
+func (l *Limiter) RequestTicket() *Ticket {
+	worker := <-l.workers
+	return &Ticket{
+		limiter: l,
+		worker:  worker,
+		dirty:   int32(0),
+	}
+}
+
+// Execute Executes a function with the work ticket
+func (t *Ticket) Execute(job func()) error {
+	if atomic.LoadInt32(&t.dirty) > 0 {
+		return fmt.Errorf("did not schedule job - ticket was dirty")
+	}
+
+	atomic.AddInt32(&t.dirty, 1)
+	atomic.AddInt32(&t.limiter.inProgress, 1)
+	job()
+
+	return nil
+}
+
+// CompleteWork signals that the execution for this particular task is complete
+func (t *Ticket) Done() error {
+	if atomic.LoadInt32(&t.dirty) > 1 {
+		return fmt.Errorf("did not complete job - job already completed")
+	}
+	// Add enough to be considered completed
+	atomic.AddInt32(&t.dirty, 2)
+	atomic.AddInt32(&t.limiter.inProgress, -1)
```

Comment:
I'm concerned when I see three atomic operations in a row, because while each read or write will be atomic with respect to each other or with respect to other calls to `Done()`, I suspect your critical section is actually the entire method. Consider this sequence of operations:

- Thread 1 and thread 2 share a reference to a `Ticket` which has `dirty = 1` and `worker = 12`
- Thread 1 calls `ticket.Done()`, gets past the atomic test on line 59, then gets preempted before it can execute line 63
- Thread 2 calls `ticket.Done()`, gets past the atomic test on line 59 (because thread 1 hasn't yet incremented `ticket.Dirty`), then continues to the end of the method, sending `12` into `ticket.limiter.workers`, then gets preempted
- Thread 1 runs again, and executes the remainder of `Done()`, also sending `12` into `ticket.limiter.workers`.

So `t.limiter.workers` will have a duplicate value sent to it, and then eventually some other call to `Done()` will block forever because there's an extra value in the `workers` channel.

I think what you need here is to add a `sync.Mutex` member to the `Limiter` and use it to protect the entire body of this method, since the whole thing is the critical section. I think that will be more natural to do if you refactor `Done()` to be a method on `Limiter` instead of `Ticket`.

Reply:
I think I can use atomic.CompareAndSwapInt32 to avoid using locks

Reply:
Why avoid locks though? This isn't performance critical and I think it'll be much, much easier to get this right with a mutex.

Reply:
I'm +1 for just using a lock.

Reply:
But what would I be locking now? Doesn't the atomic function remove the need for it?

Reply:
Atomics are harder to get right and harder to reason about, making it harder to maintain this code in the future. Note that in your most recent version, you have to add comments indicating which lines are and are not in the critical section. This would be vastly simpler if it just used a mutex to protect the entire method body, and in general we ought to aim to do the simplest thing that will solve the problem.

Reply:
Ahhh I see, fair!

Reply:
Huh actually with the changes to how tickets are done, I don't need locking or atomics at all. Technically I'm using atomics for inProgress stuff, but it seems like that's a non-critical part of the code.

Reply:
Hah, yep. I wrote a whole novel below but you summed it up in a sentence. :-D

BTW I recommend pulling out the `inProgress` stuff. You're not using it right now, and "YAGNI" applies. Also, workflow-manager only pushes metrics when it's done with everything, so even if you added metrics export for that, you'd never see a nonzero `inProgress`.

Reply:
I actually **really** enjoyed reading that - it's definitely a pattern I'm going to consider in the future. Maybe if we realize we need that in some part of workflow manager.

Concurrency is always difficult and I love reading and learning about the patterns others come up with. Going to definitely bookmark this one.